Bluetooth: Audio: Volume Control Service and client

Thalley opened the pull request:
This PR implements the Volume Control Service (VCS) server and client. 

The implementation supports and uses the Audio Input Control Service (AICS) server and client implementations as well as the Volume Offset Control Service (VOCS) server and client implementations. 

Besides the implementation, a full test suite that exercises the entire API is also added using Babblesim.

Thalley commented:
Depends on https://github.com/zephyrproject-rtos/zephyr/pull/34169 and https://github.com/zephyrproject-rtos/zephyr/pull/34127 to pass CI

aescolar commented on tests/bluetooth/bsim_bt/bsim_test_audio/test_scripts/vcs.sh:
```diff
@@ -0,0 +1,54 @@
+
+cd ${BSIM_OUT_PATH}/bin
+
+printf "\n\n======== Running VCS standalone (API) test =========\n\n"
```
Out of curiosity more than anything:
Why do you have 2 tests in the same script? (having them as 2 separate scripts allows running them in parallel, and would save you the need for the verbosity)

Thalley replied:
Mostly due to my local development process I suppose: Easier to run just a single script to test the entire VCS. 

I could split it into two scripts but then I'd probably add a third script to run both of those anyhow :D

aescolar replied:
aham :)
Maybe you'd want to use as 3rd script the one that is used in CI:
RESULTS_FILE=${ZEPHYR_BASE}/myresults.xml SEARCH_PATH=tests/bluetooth/bsim_bt/bsim_test_audio/ ./run_parallel.sh

Thalley replied:
If I'm just testing e.g. VCS locally, then I have no intention of building all and running all tests (which does take quite a while compared), so I probably wouldn't do that :D

Thalley committed:
Bluetooth: Audio: Volume Control Service and client

This commit implements the volume control service (VCS) and
client, The implementation supports and uses the
Audio Input Control Service (AICS) and
Volume Offset Control Service (VOCS) secondary services.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: BSIM test suite for VCS

Adds a Babblesim test suite for the volume control service
and client, that exercises the entire API and most of the
VCS functionality.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+ * @brief Discover Volume Control Service and included services.
+ *
+ * This will start a GATT discovery and setup handles and subscriptions.
+ * This shall be called once before any other actions can
```
Insert "be".
Before other actions can be completed, or before other actions can be started?

Thalley replied:
changed complete to `executed`

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+};
+
+/**
+ * @brief Discover Volume Control Service and included services.
```
Client side only

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+
+/**
+ * @brief Set the Volume Control Service volume step size.
+ *
```
Server side only.

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+/**
+ * @brief Set the Volume Offset Control Service location.
+ *
+ * @param conn       Connection to peer device, or NULL to read local server
```
"read" -> "set"

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+ * @brief Set the Volume Offset Control Service offset state.
+ *
+ * @param conn    Connection to peer device, or NULL to read local server value.
+ * @param inst    Pointer to the Volume Offset Control Service instance.
```
"read" -> "set"

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+/**
+ * @brief Set the Volume Offset Control Service description.
+ *
+ * @param conn          Connection to peer device, or NULL to read local server
```
"read" -> "set"

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+ * @param conn          Connection to peer device, or NULL to read local server
+ *                      value.
+ * @param inst          Pointer to the Volume Offset Control Service instance.
+ * @param description   The description to set. Value will be copied.
```
"The value"

Thalley replied:
Remove that last part as the value is no longer copied (at least not by VCS/VOCS). I assume that, unless explicitely stated, that values does not need to remain static after the call (e.g. how it's defined for gatt_subscribe etc.)

asbjornsabo replied:
Agree

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+ * @brief Deactivates a Audio Input Control Service instance.
+ *
+ * Audio Input Control Services are activated by default, but this will allow
+ * the server deactivate a Audio Input Control Service.
```
Insert "to" - "to deactivate"

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+				const char *description);
+
+/**
+ * @brief Deactivates a Audio Input Control Service instance.
```
"Deactivates" -> "Deactivate", to be consistent with the imperative form of the other functions.

"a" -> "an"

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+int bt_vcs_aics_deactivate(struct bt_aics *inst);
+
+/**
+ * @brief Activates a Audio Input Control Service instance.
```
"Activate an"

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+ * @brief Activates a Audio Input Control Service instance.
+ *
+ * Audio Input Control Services are activated by default, but this will allow
+ * the server reactivate a Audio Input Control Service instance after it has
```
"to"

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+ *
+ * @return 0 if success, errno on failure.
+ */
+int bt_vcs_aics_unmute(struct bt_conn *conn, struct bt_aics *inst);
```
Documentation says "mute", function says "unmute"

Thalley replied:
Hah, I think we had that same comment for AICS, right? 
Will fix of course :)

asbjornsabo replied:
It was familiar, yes.

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+ *
+ * @return 0 if success, errno on failure.
+ */
+int bt_vcs_aics_mute(struct bt_conn *conn, struct bt_aics *inst);
```
Documenation says "unmute", function says "mute"

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+ * @brief Set input gain to manual.
+ *
+ * @param conn   Connection to peer device, or NULL to read local server value.
+ * @param inst   Pointer to the Audio Input Control Service instance.
```
"read" -> "set"

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+ * @brief Set the input gain to automatic.
+ *
+ * @param conn   Connection to peer device, or NULL to read local server value.
+ * @param inst   Pointer to the Audio Input Control Service instance.
```
Ditto

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+/**
+ * @brief Set the input gain.
+ *
+ * @param conn   Connection to peer device, or NULL to read local server value.
```
Ditto

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+ * @brief Set the Audio Input Control Service description.
+ *
+ * @param conn          Connection to peer device, or NULL to read local server
+ *                      value.
```
Ditto

asbjornsabo commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+ * @param conn          Connection to peer device, or NULL to read local server
+ *                      value.
+ * @param inst          Pointer to the Audio Input Control Service instance.
+ * @param description   The description to set.
```
Is the description copied here as well?

Thalley replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/34314#discussion_r617449778

asbjornsabo reviewed (commented):
Comments so far.
I think the API is good, but that it must be made very clear how this works on server and client sides.  Also, some functions are one side only, and should be marked as such.

In addition, a lot of nitpicking.

Thalley commented on subsys/bluetooth/audio/Kconfig.vcs:
```diff
@@ -0,0 +1,113 @@
+config BT_VCS_AICS_INSTANCE_COUNT
+	int "Audio Input Control Service instance count for VCS"
+	default 0
+	range 0 BT_AICS_MAX_INSTANCE_COUNT
```
Correct, or even other vendor specific services. I don't think there's any way we can at compile time determine if the values are correct.

Thalley commented on subsys/bluetooth/shell/CMakeLists.txt:
```diff
@@ -25,6 +25,10 @@ zephyr_library_sources_ifdef(
+zephyr_library_sources_ifdef(
+  CONFIG_BT_VCS
+  vcs.c
+  )
```
A quick glace of the CMakelists makes the norm 
```
zephyr_library_sources_ifdef(CONFIG_XXX, YYY.c)
```
So the shell Cmakelists is the odd one out (there are a few other files with similar formatting though.

asbjornsabo replied:
OK - suggest to not fix that inconsistency here.

asbjornsabo commented on subsys/bluetooth/audio/vcs_client.c:
```diff
@@ -0,0 +1,935 @@
+					bt_vocs_client_free_instance_get();
+
+				__ASSERT(vcs_insts[i].vocs[j],
+					 "Could not allocated VOCS client instance");
```
"allocated" -> "allocate"

asbjornsabo commented on subsys/bluetooth/audio/vcs_client.c:
```diff
@@ -0,0 +1,935 @@
+
+				__ASSERT(vcs_insts[i].aics[j],
+					"Could not allocated VOCS client instance");
+
```
1) "allocated" -> "allocate"
2) "VOCS" -> "AICS"

Thalley commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+ *                     peer device.
+ */
+typedef void (*bt_vcs_discover_cb)(struct bt_conn *conn, int err,
+				   uint8_t vocs_count, uint8_t aics_count);
```
@asbjornsabo Any suggestions on how to? 

One suggestion I have, is to refactor the callback struct to be similar to `struct bt_conn_cb` where the callbacks are defined inside the struct, and "special" cases are guarded by ifdefs, so e.g. the vcs callback structure would be something like

```
struct bt_vcs_cb {
	void (*state_changed_cb)(struct bt_conn *conn, int err, uint8_t volume, uint8_t mute);
#if defined(CONFIG_BT_VCS_CLIENT)
	void (*discover_cb)(struct bt_conn *conn, int err, uint8_t vocs_count, uint8_t aics_count);
#endif /* CONFIG_BT_VCS_CLIENT */
};
```

asbjornsabo replied:
I had a more naive thought - just mention it in the Doxygen documentation.
Since the rule for most functions is that they apply on both sides, it should be sufficient to note the exceptions, e.g. like 
" * (Client side only)"

Thalley replied:
I've added that for all server or client specific functions and callbacks.

asbjornsabo commented on subsys/bluetooth/audio/vcs_client.c:
```diff
@@ -0,0 +1,935 @@
+		if (cb) {
+			if (cb->vocs_cb.discover) {
+				BT_WARN("VCS overwrote discover callback of VOCS");
+			}
```
I needed to think on this one.
The VOCS discover callback is overwritten (or set) in all cases - whether it was set before or not.
But the _warning_ is only given if it is set before.

Thalley replied:
I thought that made sense; if it was not set, then it was not really overwritten :)

asbjornsabo replied:
Agreed - it is absolutely correct.  But I somehow came from a different angle.
Maybe it could be augmented with an 
else
        BT_DEBUG("Set the discover callback of VOCS")?
Or possibly a comment of "/* Set VOCS callback - overwrite if already set */ ?

Thalley replied:
Actually, the better solution here might be to simply not allow the application to set the callback (can be verified in the register function), assuming that the application does not change the callback structure after registering (but that assumption goes for basically all callbacks in Zephyr). 

@asbjornsabo Do you think that makes more sense?

asbjornsabo replied:
I am starting to feel that I am making a lot of noise about a little thing ...
The thing works, and is correct, as it is, so although I was confused there is no pressing need for a change.
But I can not see now why the application would want to supply it's own callback, so your proposal makes sense.
Do as you think is best.

Thalley commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+ * Called when the value is read,
+ * or if the value is changed by either the server or client.
+ *
+ * @param conn    The connection to the peer device.
```
Added more text for the conn pointer. Please re-review.

asbjornsabo reviewed (commented):
Additional comments

Thalley commented on include/bluetooth/audio/vcs.h:
```diff
@@ -0,0 +1,504 @@
+ * @brief Callback function for Volume Control Service volume state.
+ *
+ * Called when the value is read,
+ * or if the value is changed by either the server or client.
```
Rewrote the text somewhat similar to your suggestion. Tried to be as concise as possibly. Please re-review.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_audio/prj.conf:
```diff
@@ -0,0 +1,51 @@
+CONFIG_BT_MAX_PAIRED=5
+
+CONFIG_BT_AUDIO=y
+CONFIG_BT_AUDIO_UNICAST=y
```
CONFIG_BT_AUDIO_UNICAST isn't will remove

Thalley commented on subsys/bluetooth/audio/vcs_client.c:
```diff
@@ -0,0 +1,935 @@
+					 const void *data, uint16_t length)
+{
+	uint16_t handle = params->value_handle;
+	struct vcs_instance *vcs_inst = &vcs_insts[bt_conn_index(conn)];
```
I did consider adding an __ASSERT for this, but no:
The connection indexes are static and vcs_insts are static as well, so regardless of the conn (unless conn is NULL), then vcs_inst will always be non-null.

asbjornsabo replied:
What happens if the devices is connected to devices A, B and C, with connection indexes 0, 1 and 2, but only A and C has VCS services?

Thalley replied:
Then 1) this (notify) callback will never be called for the connection between the device and B (since it never subscribed), and even if that happens (i.e. the server notifies with a wrong handle or something similar to that), then the `vcs_inst` pointer is still a valid pointer, but where the callbacks are simply not registered.

asbjornsabo commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_client_test.c:
```diff
@@ -0,0 +1,730 @@
+	}
+
+	strncpy(g_vocs_desc, description, sizeof(g_vocs_desc) - 1);
+
```
Set terminating zero?

Thalley replied:
`strncpy` should handle that for us.

asbjornsabo replied:
"The  strncpy()  function  is  similar,  except that at most n bytes of src are copied.  Warning: If there is no null byte among the first n bytes of src, the string placed in dest will not be null-terminated."
(This is from the Linux manpage, but I expect zephyr is the same.)

The code checks against "sizeof() - 1", and also copies that many bytes.  I expected the last byte (the -1) to be used for the termination (in case of truncation), but it isn't.

asbjornsabo commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_client_test.c:
```diff
@@ -0,0 +1,730 @@
+	}
+
+	strncpy(g_aics_desc, description, sizeof(g_aics_desc) - 1);
+
```
Termination?

Thalley replied:
`strncpy` should handle that for us.

asbjornsabo commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_client_test.c:
```diff
@@ -0,0 +1,730 @@
+		return;
+	}
+	WAIT_FOR(g_mute == expected_mute && g_cb && g_write_complete);
+	printk("VCS volume upped\n");
```
VCS muted

asbjornsabo commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_test.c:
```diff
@@ -0,0 +1,754 @@
+	}
+
+	strcpy(g_vocs_desc, description);
+
```
In the client test, there is a size check.
(Not sure how important that is here.)

Thalley replied:
Changed to use `strncpy`, but should not be necessary as we use the Kconfig that will have truncated the value from the service implementation anyhow :)

asbjornsabo reviewed (commented):
Last part of first round - mostly skimmed quickly through the shell commands, though.

Thalley commented on subsys/bluetooth/audio/vcs.c:
```diff
@@ -0,0 +1,910 @@
+			return true;
+		}
+	}
+#endif /* CONFIG_BT_VCS */
```
The `#if defined(CONFIG_BT_VCS) || defined(CONFIG_BT_VCS_CLIENT)` is superfluous as vcs.c is only compiled if one of those are true anyhow. 

I'm not sure what else you mean by your comment.

asbjornsabo replied:
The #if that is removed, and the remaining `#if defined(CONFIG_BT_VCS)` seemed to imply that this function could also be called for the client. But I don't think it is?

Thalley commented on subsys/bluetooth/shell/vcs.c:
```diff
@@ -0,0 +1,824 @@
+{
+	int result;
+	int step = strtol(argv[1], NULL, 0);
+
```
Yes, I'm not sure how much we want to care about this to be honest. It doesn't seem to be done "properly" for most cases, but the return value is even a `long` and not an `int`. 

So yeah, ideally we would take all the results an store in a `long` value, check the value range and then use the value.

asbjornsabo replied:
I think we do not care.

asbjornsabo reviewed (approved):
All my comments have been resolved to my satisfaction.

carlescufi commented:
@Vudentz please review before Friday so we can get this into Zephyr 2.6.

Thalley commented:
@carlescufi We also need to resolve https://github.com/zephyrproject-rtos/zephyr/issues/34823 before this can be merged, as the MISRA scripts are blocking this PR (and others).

jhedberg commented:
I don't mind so big changes (6000+ lines added) being in a single pull request, but you could IMO do a bit better job at splitting things up into smaller self-contained commits. E.g. the service and client could probably be in separate commits. I only took a quick look at the code itself (too much for a thorough review) and that looks fairly alright to me.

asbjornsabo commented:
> I don't mind so big changes (6000+ lines added) being in a single pull request, but you could IMO do a bit better job at splitting things up into smaller self-contained commits. E.g. the service and client could probably be in separate commits. I only took a quick look at the code itself (too much for a thorough review) and that looks fairly alright to me.

For what it is worth, before upstreaming the code has been developed and reviewed in smaller portions, as a number of changes over time, in the zephyr-bt-sig tree.

Thalley commented:
> I don't mind so big changes (6000+ lines added) being in a single pull request, but you could IMO do a bit better job at splitting things up into smaller self-contained commits. E.g. the service and client could probably be in separate commits. I only took a quick look at the code itself (too much for a thorough review) and that looks fairly alright to me.

Yes, Luiz also gave that comment. Splitting the server and client is somewhat difficult as they share the same API, i.e. the vcs.h header file would be hard to split up into two commits. The alternative is just add e.g. the server and vcs.h (with references to the client) in one commit, and then the client in another, but that would be fairly weird. Perhaps I could also add the server and client in two commits, and then add the public header file. 

What I've done for the MICS PR is that I've split it into 3: 
1) Implementation
2) Shell
3) Tests

carlescufi commented:
> @carlescufi We also need to resolve #34823 before this can be merged, as the MISRA scripts are blocking this PR (and others).

We have been overriding that in other PRs, and we can do that here too.

Thalley commented:
> > @carlescufi We also need to resolve #34823 before this can be merged, as the MISRA scripts are blocking this PR (and others).
> 
> We have been overriding that in other PRs, and we can do that here too.

@carlescufi I think this is good to merge now. Can you override the MISRA script?
